### PR TITLE
Extensions example

### PR DESCRIPTION
**Pull request by iNikem**
No more multi-project gradle build.

### CODE REVIEW

File: examples/extension/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+There are several options to override or customise instrumentation provided by the upstream agent.
+The following description follows one specific use-case:
+
+> Instrumentation X from Otel distribution creates span that I don't like and I want to change it in my vendor distro.
+
+As an example, let us take some database client instrumentation that creates a span for database call
+and extracts data from db connection to provide attributes for that span.
+
+### I don't want this span at all
+The easiest case. You can just pre-configure your distribution and disable given instrumentation.
```
Comment (mateuszrzeszutek) [inline examples/extension/README.md:53]: > pre-configure your distribution

It's not really a distribution anymore, is it? Maybe something like 
```suggestion
The easiest case. You can just pre-configure the agent in your extension and disable given instrumentation.
```
would be more accurate?

File: examples/extension/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+### I want to add/modify some attributes and their values does NOT depend on a specific db connection instance.
+E.g. you want to add some data from call stack as span attribute. 
+In this case just provide your custom `SpanProcessor`.
+No need for touching instrumentation itself.
+
+### I want to add/modify some attributes and their values depend on a specific db connection instance.
+Write a _new_ instrumentation which injects its own advice into the same method as the original one.
+Use `getOrder` method to ensure it is run after the original instrumentation.
```
Comment (mateuszrzeszutek) [inline examples/extension/README.md:62]:
```suggestion
Use `order` method to ensure it is run after the original instrumentation.
```

File: examples/extension/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Now you can augment current span with new information.
+
+See [DemoServlet3Instrumentation](instrumentation/servlet-3/src/main/java/com/example/javaagent/instrumentation/DemoServlet3Instrumentation.java).
```
Comment (mateuszrzeszutek) [inline examples/extension/README.md:65]:
```suggestion
See [DemoServlet3InstrumentationModule](instrumentation/servlet-3/src/main/java/com/example/javaagent/instrumentation/DemoServlet3InstrumentationModule.java).
```

File: examples/extension/build.gradle [hunk 4]
```diff
@@ -1,52 +1,115 @@
...
+  compileOnly("io.opentelemetry.javaagent:opentelemetry-javaagent-extension-api:${versions.opentelemetryJavaagentAlpha}")
+
+  //Provides @AutoService annotation that makes registration of our SPI implementations much easier
+  compileOnly deps.autoservice
+  annotationProcessor deps.autoservice
 
-    compileJava {
-      options.release.set(11)
+  /*
+   Used by our instrumentation module to references classes of the target instrumented library.
```
Comment (mateuszrzeszutek) [inline examples/extension/build.gradle:62]: Nit:
```suggestion
   Used by our demo instrumentation module to references classes of the target instrumented library.
```

**Review by mateuszrzeszutek — APPROVED** [review]

File: examples/extension/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+* [DemoSpanProcessor](src/main/java/com/example/javaagent/DemoSpanProcessor.java) - custom `SpanProcessor`
+* [DemoSpanExporter](src/main/java/com/example/javaagent/DemoSpanExporter.java) - custom `SpanExporter`
+* [DemoServlet3InstrumentationModule](src/main/java/com/example/javaagent/instrumentation/DemoServlet3InstrumentationModule.java) - additional instrumentation
+
+## Instrumentation customisation
+
+There are several options to override or customise instrumentation provided by the upstream agent.
+The following description follows one specific use-case:
+
+> Instrumentation X from Otel distribution creates span that I don't like and I want to change it in my vendor distro.
```
Comment (trask) [inline examples/extension/README.md:47]:
```suggestion
> Instrumentation X from Otel distribution creates span that I don't like and I want to change it.
```

File: examples/extension/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+### I want to remove some attributes
+Write custom exporter or use attribute filtering functionality in Collector.
+
+### I don't like Otel span at all. I want to significantly modify it and its lifecycle
+Disable existing instrumentation.
+Write a new one, which injects `Advice` into the same (or better) method as the original instrumentation.
+Write your own `Advice` for this.
+Use existing `Tracer` directly or extend it.
+As you have your own `Advice`, you can control which `Tracer` you use.
```
Comment (trask) [inline examples/extension/README.md:75]: migrate ref to `Tracer` now or later?

File: examples/extension/build.gradle [hunk 7]
```diff
@@ -1,52 +1,114 @@
+plugins {
+  id "java"
+
+  /*
+  Instrumentation agent extension mechanism expects a single jar containing everything required
+  for your extension. This also includes any external libraries that your extension uses and
+  cannot access from application classpath (see comment below about `javax.servlet-api` dependency).
+
+  Thus we use Shadow Gradle plugin to package our classes and all required runtime dependencies
+  into a single fat jar.
```
Comment (trask) [inline examples/extension/build.gradle:10]:
```suggestion
  into a single jar.
```

File: examples/extension/build.gradle [hunk 8]
```diff
@@ -1,52 +1,114 @@
...
+  compileOnly("io.opentelemetry.javaagent:opentelemetry-javaagent-extension-api:${versions.opentelemetryJavaagentAlpha}")
+
+  //Provides @AutoService annotation that makes registration of our SPI implementations much easier
+  compileOnly deps.autoservice
+  annotationProcessor deps.autoservice
 
-    compileJava {
-      options.release.set(11)
+  /*
+   Used by our demo instrumentation module to references classes of the target instrumented library.
```
Comment (trask) [inline examples/extension/build.gradle:62]:
```suggestion
   Used by our demo instrumentation module to reference classes of the target instrumented library.
```

File: examples/extension/build.gradle [hunk 9]
```diff
@@ -1,52 +1,114 @@
...
+   We again use `compileOnly` here because during runtime these classes are provided by the
+   actual application that we instrument.
+
+   NB! Only Advice classes of instrumentation modules can access classes from application classpath.
```
Comment (trask) [inline examples/extension/build.gradle:66]:
```suggestion
   NB! Only Advice (and "helper") classes of instrumentation modules can access classes from application classpath.
```

File: examples/extension/build.gradle [hunk 10]
```diff
@@ -1,52 +1,114 @@
...
+   See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#advice-classes
+   */
+  compileOnly group: 'javax.servlet', name: 'javax.servlet-api', version: '3.0.1'
+
+  /*
+  This dependency is required DemoSpanProcessor both during compile and runtime.
```
Comment (trask) [inline examples/extension/build.gradle:72]:
```suggestion
  This dependency is required for DemoSpanProcessor both during compile and runtime.
```

File: examples/extension/build.gradle [hunk 11]
```diff
@@ -1,52 +1,114 @@
...
+  Only dependencies added to `implementation` configuration will be picked up by Shadow plugin
+  and added to the resulting fat jar for our extension's distribution.
```
Comment (trask) [inline examples/extension/build.gradle:74]:
```suggestion
  and added to the resulting jar for our extension's distribution.
```

File: examples/extension/build.gradle [hunk 12]
```diff
@@ -1,52 +1,114 @@
...
+
+tasks {
+  test {
+    useJUnitPlatform()
+
+    def extensionJar = tasks.shadowJar
+    inputs.files(layout.files(extensionJar))
+
+    doFirst {
+      //To run our application with javaagent published by OpenTelemetry Java instrumentation project
```
Comment (trask) [inline examples/extension/build.gradle:102]:
```suggestion
      //To run our tests with the javaagent published by OpenTelemetry Java instrumentation project
```

**Review by trask — APPROVED** [review]
❤️

File: examples/extension/README.md (hunk 6 above)
Comment (iNikem) [reply examples/extension/README.md:75]: To `Instrumenter`? I vote for later.

**Pull request merged**